djangosaml2idp/models.py

```diff
@@ -51,26 +51,56 @@ def refresh_metadata(self, force_refresh: bool = False) -> bool:
-            if self.remote_metadata_url:
-                try:
-                    self.local_metadata = validate_metadata(fetch_metadata(self.remote_metadata_url))
-                    # TODO: moving this line here passed the test.
-                    self.metadata_expiration_dt = extract_validuntil_from_metadata(self.local_metadata)
-                except Exception:
-                    # TODO: should we return False?
-                    logger.error(f'Metadata for SP {self.entity_id} could not be pulled from remote url {self.remote_metadata_url}.')
-            elif self.metadata_expiration_dt and now() > self.metadata_expiration_dt:
-                # TODO: should we also return False here?
-                logger.error(f'Metadata for SP {self.entity_id} has expired, no remote metadata found to refresh.')
-
-            # TODO: commented this out to pass test
-            # self.metadata_expiration_dt = extract_validuntil_from_metadata(self.local_metadata)
-            return True
-        return False
+        # Refresh in these cases:
+        should_refresh = False
+        # - Data was not fetched ever before, so local_metadata is empty
+        if not self.local_metadata:
+            should_refresh = True
+        # - The expiration timestamp is not set or expired
+        if not self.metadata_expiration_dt or now() > self.metadata_expiration_dt:
+            should_refresh = True
+        # - Force refresh requested by caller
+        if force_refresh:
+            should_refresh = True
+
+        if not should_refresh:
+            return False
+
+        # Try to do the refresh if a remote is set
+        if self.remote_metadata_url:
+            try:
+                self.local_metadata = validate_metadata(fetch_metadata(self.remote_metadata_url))
+                self.metadata_expiration_dt = extract_validuntil_from_metadata(self.local_metadata).replace(tzinfo=None)
+                # Return True if it is now valid, False (+ log an error) otherwise
+                if now() > self.metadata_expiration_dt:
+                    logger.error(f'Remote metadata for SP {self.entity_id} was refreshed, but contains an expired validity datetime.')
+                    return False
+                return True
+            except Exception as e:
+                logger.error(f'Metadata for SP {self.entity_id} could not be pulled from remote url {self.remote_metadata_url}.', extra={'exception': str(e)})
+                return False
+
+        # No remote metadata url to fetch new data from. Log an error anyways in case
+        # - the current local data is not present
+        # - if it does not have a valid expiry timestamp
+        # - if it is expired
+        if not self.local_metadata:
+            logger.error(f'Local metadata for SP {self.entity_id} is not present, and no remote metadata found to refresh.')
+            return False
```

Comment:
I think this case should be moved to after `if not should_refresh` as it is a short circuit case. The if would change to. `if not self.remote_metadata_url and not self.local_metadata` that would make it more obvious what the case is. Right now you would have to go back and read through all the cases under `if self.remote_metadata_url` and see that they all return to realize how you get to this case and that `if self.remote_metadata_url` has to be falsy to get here. I just think it might be more clear to explicitly check both.

```python
if not should_refresh:
    return False

if not self.remote_metadata_url and not self.local_metadata:
    return False

if self.remote_metadata_url:
    return self._refresh_from_remote()

if (not self.metadata_expiration_dt) or (now() > self.metadata_expiration_dt):
    return self._refresh_from_local()
```

Reply:
And even further refactor could be

```python
def refresh_metadata(self, force_refresh):
    if not force_refresh and self._should_refresh() is False:
        return False
    
    if self.remote_metadata_url:
        return self._refresh_from_remote()

    if (not self.metadata_expiration_dt) or (now() > self.metadata_expiration_dt):
        return self._refresh_from_local()
```
I'm not sure about the `raise Exception` at the end but I haven't quite thought through what to do. We SHOULD be able to know all of the possible cases so it should be unnecessary to have it. We should be able to write a test case for it and then decide what to actually do in that event.

Reply:
I've split it up in methods due to your suggestion. I've left the exception at the end for now though; while ideally we do cover all cases and have them formally tested, raising an exception here will alert users and us in case we don't, which is quite plausible for now. Then we can investigate and discover a case which we haven't covered yet, should we silently fail and e.g. just return False, it might lead to unexpected behaviour. Should this never pop up in the future, it can be dropped as then there's a bit more confidence in having covered all cases.